pkg/secrets/json_expander.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package secrets
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/bitly/go-simplejson"
+)
+
+type JSONExpander struct {
+	sm SecretManager
+}
+
+// NewJSONExpander creates a new secret manager that allows secret values to be stored as json.
+// When resolving secrets, if "dot-notation" is provided via the secret-name, the secret value
+// will be json-decoded and the dot-notation will be used to resolve the secret value.
+func NewJSONExpander(ctx context.Context, f SecretManagerFunc) (SecretManager, error) {
+	sm, err := f(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new JSONExpander SecretManager: %w", err)
+	}
+
+	return WrapJSONExpander(ctx, sm)
+}
+
+// WrapJSONExpander wraps an existing SecretManager with json-expansion logic.
+func WrapJSONExpander(ctx context.Context, sm SecretManager) (SecretManager, error) {
+	return &JSONExpander{
+		sm: sm,
+	}, nil
+}
+
+// GetSecretValue implements the SecretManager interface, but allows for json-expansion
+// of the secret-value. If the secret name contains a period, the secret value is expected
+// to be json. The secret name is assumed to come before the period, while the map-key is
+// expected to follow.
+//
+// For example:
+// If a secret with a name of "psqlcreds" has a value of `{"username":"gandalf", "password":"abc"}`
+// When GetSecretValue(ctx, "psqlcreds") is called, the raw json value will be returned.
+// When GetSecretValue(ctx, "psql.username") is called, only "gandalf" (without quotes) will be returned.
+func (sm *JSONExpander) GetSecretValue(ctx context.Context, name string) (string, error) {
+	parts := strings.Split(name, ".")
+	if len(parts) == 1 {
+		return sm.sm.GetSecretValue(ctx, name)
+	}
+
+	if len(parts) > 2 {
+		return "", fmt.Errorf("secret-name has more than one period - json expansion path undeterminable")
+	}
+
+	// Get the value from the embdedded secret-manager using what comes before the period.
+	val, err := sm.sm.GetSecretValue(ctx, parts[0])
+	if err != nil {
+		return "", err
+	}
+
+	json, err := simplejson.NewJson([]byte(val))
```

Comment:
Could we  unmarshal into a `map[string]string` (or `map[string]interface{}`) instead of depending on a library? It'll be a few extra LOC, but isn't super complex. We could also move it into a helper for easier testing.

Reply:
Done. Thanks for the suggestion. I like the added ability to have nested values!